### PR TITLE
Hiding attributes when optional and no value is defined

### LINKED ISSUE #121: empty and optional elements appear in json body
The following blueprint:

```md
# GET /
+ Response 200 (application/json)
    + Attributes
        + s1 (optional)
```
Renders the following JSON example

```json
"content": "{\n  \"s1\": \"\"\n}"
```

`s1` should not be part of the body, because it is both empty and optional.

### PR DESCRIPTION
**Pull request by obihann**
This resolves #121 by hiding attributes from the JSON body when they are both empty and optional.

### CODE REVIEW

File: src/refract/RenderJSONVisitor.cc [hunk 1]
```diff
@@ -170,7 +199,6 @@ namespace refract
             if (element.empty() && isNullable(element)) {
                 return NULL;
             }
-
```
Comment (pksunkara) [inline src/refract/RenderJSONVisitor.cc]: Why is this line removed?

File: src/refract/RenderJSONVisitor.cc [hunk 2]
```diff
@@ -89,6 +89,33 @@ namespace refract
 
         return false;
     }
+    
+    template<typename T>
+    bool isOptional(const T& e) {
```
Comment (pksunkara) [inline src/refract/RenderJSONVisitor.cc]: This function is just duplicate of `isNullable`. Please optimise it by making a single function named `isTypeAttribute` taking a single argument.

**Comment by pksunkara** [comment]
Review done. Please do the proposed changes.

File: src/refract/RenderJSONVisitor.cc (hunk 1 above)
Comment (obihann) [inline src/refract/RenderJSONVisitor.cc]: A little over-zealous cleaning on my part, I'll fix that up now.

File: src/refract/RenderJSONVisitor.cc (hunk 2 above)
Comment (obihann) [inline src/refract/RenderJSONVisitor.cc]: Sounds good, I'll do that now.

File: src/refract/RenderJSONVisitor.cc [hunk 3]
```diff
@@ -8,6 +8,7 @@
 
 #include "Element.h"
 #include "Visitors.h"
+#include "../Serialize.h"
```
Comment (pksunkara) [inline src/refract/RenderJSONVisitor.cc]: Why is this here?

File: src/refract/RenderJSONVisitor.cc [hunk 4]
```diff
@@ -94,8 +95,10 @@ namespace refract
         RenderJSONVisitor renderer;
 
         if (e.value.second) {
-            if (isNullable(e) && e.value.second->empty()) {
+            if (IsTypeAttribute(e, drafter::SerializeKey::Nullable) && e.value.second->empty()) {
```
Comment (pksunkara) [inline src/refract/RenderJSONVisitor.cc]: Please use `"nullable"`

File: src/refract/RenderJSONVisitor.cc [hunk 5]
```diff
@@ -167,10 +170,10 @@ namespace refract
                 return &d->value;
             }
 
-            if (element.empty() && isNullable(element)) {
+            if (element.empty() && IsTypeAttribute(element, drafter::SerializeKey::Nullable)) {
```
Comment (pksunkara) [inline src/refract/RenderJSONVisitor.cc]: Please use `"nullable"`

File: src/refract/RenderJSONVisitor.cc [hunk 6]
```diff
@@ -94,8 +95,10 @@ namespace refract
         RenderJSONVisitor renderer;
 
         if (e.value.second) {
-            if (isNullable(e) && e.value.second->empty()) {
+            if (IsTypeAttribute(e, drafter::SerializeKey::Nullable) && e.value.second->empty()) {
                 renderer.result = sos::Null();
+            } else if (IsTypeAttribute(e, drafter::SerializeKey::Optional) && e.value.second->empty()) {
```
Comment (pksunkara) [inline src/refract/RenderJSONVisitor.cc]: Please use `"optional"`

**Comment by pksunkara** [comment]
Please don't any files from drafter in refract folder. Please do the proposed changes and I can merge this.

File: src/refract/RenderJSONVisitor.cc (hunk 3 above)
Comment (obihann) [inline src/refract/RenderJSONVisitor.cc]: I included that so I could do something like `drafter::SerializeKey::Nullable` instead of `"nullable"`, however based on your other comments I'll just switch to the plain string.

**Comment by obihann** [comment]
@pksunkara Changes have been made and I rebased from master so that it is all up to date.

**Comment by pksunkara** [comment]
You didn't succeed in rebasing this properly. Please go back to how it was. There will be no merge conflicts so you need not have rebased. :disappointed:

**Comment by obihann** [comment]
@pksunkara understood. I'll revert. I thought I had to because github told me it was out of sync with master.

**Comment by pksunkara** [comment]
Your git history is all kinds of messed up. :disappointed: and this PR is messed up too.

Please make a PR with only this commit, https://github.com/obihann/drafter/commit/a5ef117c1425f7d047ae6060b5e3d9b6aacb2d26

**Comment by obihann** [comment]
Will do.
